### PR TITLE
[WIP] Mail folder widget

### PR DESCRIPTION
**Pull request by Denis1990**

### CODE REVIEW

**Comment by cketti** [comment]
As far as I can see there's no necessity for this to live inside K-9 Mail. It might as well be a stand-alone app providing the widget.

**Comment by Valodim** [comment]
I dunno, I can think of many reasons to have it, and few to keep it external. There is little in the way of interchangability other than design details where this could benefit from choice of app, the maintenance isn't going to be huge (we need to keep MessageProvider up to date anyways, and it hasn't changed in years), aaand gmail has it :)

**Comment by Denis1990** [comment]
I thinks that it would be a useful feature to have in an email app.
As for the support i intend to stick around and maintain the code.

File: k9mail/src/main/res/layout/mail_list_view_widget_layout.xml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            android:layout_height="wrap_content"
+            android:orientation="horizontal">
+        <TextView android:id="@+id/folder"
+                  android:layout_width="wrap_content"
+                  android:layout_height="wrap_content"
+                  android:textSize="20sp"
+                  android:layout_alignParentTop="true"
+                  android:layout_alignBaseline="@+id/new_message"/>
+
+        <Button android:id="@+id/new_message"
```
Comment (cketti) [inline k9mail/src/main/res/layout/mail_list_view_widget_layout.xml]: Why not use an `ImageButton`?

File: k9mail/src/main/java/com/fsck/k9/service/MailListViewWidgetService.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        mailItems = new ArrayList<>(25);
+        Cursor cursor = mContext.getContentResolver().query(
+                MessageProvider.CONTENT_URI.buildUpon().appendPath("inbox_messages").build(),
+                MAIL_LIST_PROJECTIONS,
+                null,
+                null,
+                MessageProvider.MessageColumns.SEND_DATE + "DESC");
+        while (cursor.moveToNext()) {
+            mailItems.add(new MailItem(
+                        cursor.getString(0), /* id */
```
Comment (Denis1990) [inline k9mail/src/main/java/com/fsck/k9/service/MailListViewWidgetService.java]: My concern here was that i would create 8 local variable at every iteration thereby adding a (slight albeit) memory overhead. That's why i added the comments.

Comment (cketti) [inline k9mail/src/main/java/com/fsck/k9/service/MailListViewWidgetService.java]: Readable code trumps unproven micro-optimizations. In this case the generated code is most likely the same whether you have local variables or not.

File: k9mail/src/main/java/com/fsck/k9/provider/MailListViewWidgetProvider.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        views.setTextViewText(R.id.folder, "INBOX");
+        Intent intent = new Intent(context, MailListViewWidgetService.class)
+                .putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
+        intent.setData(Uri.parse(intent.toUri(Intent.URI_INTENT_SCHEME)));
+        views.setRemoteAdapter(R.id.listView, intent);
+
+        Intent clickIntent = new Intent(context, MailListViewWidgetProvider.class);
+        clickIntent.setPackage(PACKAGE_NAME);
+        clickIntent.setAction(ACTION_VIEW_MAIL_ITEM);
+        PendingIntent pendingIntent = PendingIntent.getBroadcast(context, MessageList.REQUEST_MASK_PENDING_INTENT,
```
Comment (Denis1990) [inline k9mail/src/main/java/com/fsck/k9/provider/MailListViewWidgetProvider.java]: I've tried using PendingIntent#getActivity() method as well but nothing happened no action what so ever when i clicked on the widget.
This is due to the way you set up behaviour for individual items in a widget's collection.[more here](https://developer.android.com/guide/topics/appwidgets/index.html#implementing_collections)
The main points are that
* You need to set a pending intent template for the entire collection
* add fill intent for each item (in getViewAt method), the fill in intent is combined with the pendingIntent 
* process the event in the onReceive method of AppWidgetProvider class.

**Comment by Denis1990** [comment]
@cketti the changes are implemented.

**Comment by cketti** [comment]
I cleaned up your code a bit and opened a new pull request. See #2255

**Pull request closed without merging**